tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,259 @@
+
+  Args:
+    uri: The input uri for the specific split of the input example artifact.
+    label: The name of the column containing class names to sample by.
+    shards: The number of files that each sampled split should
+    contain. Default 0 is Beam's tfrecordio function's default.
+    null_classes: A list determining which classes that we should
+    not sample. Defaults to None.
+    output_dir: The output directory for the split of the output artifact.
+  Returns:
+    None
+  """
+
+  data = read_tfexamples(p, uri, label)
+  merged = sample_examples(data, null_classes, sampling_strategy)
+  write_tfexamples(merged, shards, output_dir)
+
+
+def read_tfexamples(p, uri, label):
+  """Function that reads tf.Examples from tfRecord files and converts them
+  to a K-V PCollection usable by Beam."""
+
+  dataset = tf.data.TFRecordDataset(tf.data.Dataset.list_files(f'{uri}/*'),
+                                    compression_type="GZIP")
+
+  # Take the input TFRecordDataset and extract the class label that we want.
+  # Output format is a K-V PCollection: {class_label: TFRecord in string format}
+  data = (p
+          | "DatasetToPCollection" >> beam.Create(dataset)
+          | "MapToLabel" >> beam.Map(generate_elements, label))
+  return data
+
+
+def sample_examples(data, null_classes, sampling_strategy):
+  """Function that performs the sampling given a label-mapped dataset."""
+
+  # Finds the minimum frequency of all classes in the input label.
+  # Output is a singleton PCollection with the minimum # of examples.
+
+  def find_minimum(elements):
+    return min(elements or [0])
+
+  def find_maximum(elements):
+    return max(elements or [0])
+
+  if sampling_strategy == spec.SamplingStrategy.UNDERSAMPLE:
+    sample_fn = find_minimum
+  else:
+    sample_fn = find_maximum
+
+  val = (data
+         | "CountPerKey" >> beam.combiners.Count.PerKey()
+         | "FilterNullCount" >>
+         beam.Filter(lambda x: filter_null(x, null_vals=null_classes))
+         | "Values" >> beam.Values()
+         | "GetSample" >> beam.CombineGlobally(sample_fn))
+
+  # Actually performs the undersampling functionality.
+  # Output format is a K-V PCollection: {class_label: TFRecord in string format}
+  res = (data
+         | "GroupBylabel" >> beam.GroupByKey()
+         | "FilterNull" >>
+         beam.Filter(lambda x: filter_null(x, null_vals=null_classes))
+         | "Sample" >> beam.FlatMapTuple(sample_data,
+                                         sampling_strategy=sampling_strategy,
+                                         side=beam.pvalue.AsSingleton(val)))
+
+  # Take out all the null values from the beginning and put them back in the pipeline
+  null = (data
+          | "ExtractNull" >> beam.Filter(
+              lambda x: filter_null(x, keep_null=True, null_vals=null_classes))
+          | "NullValues" >> beam.Values())
+  merged = (res, null) | "Merge PCollections" >> beam.Flatten()
```

Comment:
Consider just `return (res, null) | "Merge PCollections" >> beam.Flatten()` unless you feel strongly that naming this has value for the reader